### PR TITLE
registry-client: retry on transient network fails

### PR DESCRIPTION
**Pull request by isaacs**
This fixes an issue where we'd crash when the network is flaky, with `ECONNREFUSED`, `ECONNRESET`, `EPIPE`, and `UND_ERR_SOCKET`.

Using undici's `RetryHandler` takes care of this transparently, including retrying `GET` requests with a `Range` header (for servers that advertise support for it) if the body had already been partly consumed.

### CODE REVIEW

File: src/registry-client/src/index.ts [hunk 1]
```diff
@@ -91,19 +102,41 @@ const agentOptions: Agent.Options = {
...
-    cache = xdg.cache('registry-client'),
-  }: RegistryClientOptions) {
+  constructor(options: RegistryClientOptions) {
+    const {
+      cache = xdg.cache('registry-client'),
+      'fetch-retry-factor': timeoutFactor = 2,
+      'fetch-retry-mintimeout': minTimeout = 0,
+      'fetch-retry-maxtimeout': maxTimeout = 30_000,
+      'fetch-retries': maxRetries = 3,
+    } = options
```
Comment (lukekarrys) [inline src/registry-client/src/index.ts:115]: Generally, I'm not a fan of doubling up on these defaults in both the CLI and the workspace. It's already a pattern we use everywhere so I'm ok with adding another instance here.

I'd rather have our workspaces be less ergonomic and rely on types to force setting the options. I think this pattern is less likely to lead to bugs in the future if CLI defaults get changed or removed.

Again, not blocking this PR but curious to pros to this pattern that I might not be seeing.

**Review by lukekarrys — APPROVED** [review]

File: src/registry-client/src/index.ts (hunk 1 above)
Comment (isaacs) [reply src/registry-client/src/index.ts:115]: Good call. Updated, PTAL!

File: src/vlt/src/config/index.ts [hunk 2]
```diff
@@ -321,7 +317,7 @@ export class Config {
     if (this.command) p.positionals.shift()
     else this.command = getCommand(p.values['fallback-command'])
 
-    return Object.assign(this, p)
+    return Object.assign(this, p) as this & ParsedConfig
```
[the remark this answers is not in the document]
Comment (lukekarrys) [reply src/vlt/src/config/index.ts:320]: I thought about adding a code like `ETHISSHOULDNEVERHAPPEN` where the error message would be "we messed up please file a bug".

**Review by lukekarrys — APPROVED** [review]

File: src/vlt/src/config/index.ts [hunk 3]
```diff
@@ -321,7 +317,13 @@ export class Config {
     if (this.command) p.positionals.shift()
     else this.command = getCommand(p.values['fallback-command'])
 
-    return Object.assign(this, p)
+    Object.assign(this, p)
+
+    /* c8 ignore start - unpossible */
+    if (!isParsed(this)) throw error('failed to parse config')
+    /* c8 ignore stop */
```
Comment (lukekarrys) [inline src/vlt/src/config/index.ts:324]:
```suggestion
    /* c8 ignore next - unpossible */
    if (!isParsed(this)) throw error('failed to parse config')
```

File: src/vlt/src/config/index.ts [hunk 4]
```diff
@@ -321,7 +317,13 @@ export class Config {
...
+
+    return this as this & ParsedConfig
```
Comment (lukekarrys) [inline src/vlt/src/config/index.ts:326]:
```suggestion
    return this
```

**Pull request merged**